User Agent Automation and WebDriver test support

mohamedamir opened the pull request:
Closes #357

The following tasks have been completed:

- [X] Modified Web platform tests https://github.com/web-platform-tests/wpt/pull/58587

Implementation commitment:

- [x] [WebKit](https://bugs.webkit.org/show_bug.cgi?id=306292)
- [X] [Chromium](https://crbug.com/447093830) 
- [ ] Gecko (link to issue)

Documentation and checks

- [ ] Affects privacy
- [ ] Affects security
- [ ] Pinged MDN
- [ ] Updated Explainer
- [ ] Updated digitalcredentials.dev


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c-fedid/digital-credentials/pull/381.html" title="Last updated on May 11, 2026, 5:09 PM UTC (5c51691)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-fedid/digital-credentials/381/9e4579c...5c51691.html" title="Last updated on May 11, 2026, 5:09 PM UTC (5c51691)">Diff</a>

mohamedamir committed:
user agent automation section and update the algorithms

mohamedamir committed:
user agent automation section and update the algorithms

mohamedamir committed:
define "next wallet behavior"

mohamedamir committed:
define "next wallet behavior"

mohamedamir committed:
tidy

mohamedamir committed:
tidy

TallTed commented on index.html:
```diff
@@ -856,6 +856,22 @@ <h3>
+          </li>
+        </ol>
+      </li>
+      <li>Otherwise (if [=next wallet behavior=] is not set), proceed with the
```
```suggestion
      <li>Otherwise (i.e., if [=next wallet behavior=] is not set), proceed with the
```

marcoscaceres commented on index.html:
```diff
@@ -856,6 +856,22 @@ <h3>
       </li>
+      <li>If the session's [=next wallet behavior=] is set:
+        <ol>
+          <li>If the mode is "respond", resolve the promise with a new
```
Nit: let's link these up properly to their definitions (e.g. {{DigitalCredential}})... "respond" should be formally defined, etc.

marcoscaceres commented on index.html:
```diff
@@ -856,6 +856,22 @@ <h3>
+          <li>If the mode is "wait", the promise never resolves.
+          </li>
+          <li>If the mode is "decline", reject the promise with a
+          "NotAllowedError" `DOMException`.
```
```suggestion
          {{"NotAllowedError"}} {{DOMException}}.
```

mohamedamir replied:
I will revert all changes to the algorithm as discussed in another comment thread.

marcoscaceres commented on index.html:
```diff
@@ -917,6 +933,22 @@ <h3>
           </li>
         </ol>
       </li>
+      <li>If the session's [=next wallet behavior=] is set:
```
As this is duplicative, maybe move the step to its own algorithm? then we only need to maintain one set of steps.

marcoscaceres commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+          </tbody>
+        </table>
+        <p>
+          When <a data-cite="webdriver#dfn-validate-capabilities">validating
```
nit: avoid using data-cite... if something hasn't been exported from Web Driver, we should send a PR to web driver to export the definition we need.

marcoscaceres commented on index.html:
```diff
@@ -856,6 +856,22 @@ <h3>
           </li>
         </ol>
       </li>
+      <li>If the session's [=next wallet behavior=] is set:
```
I wonder if we need this these steps at all?... this should really be covered by #306.

marcoscaceres reviewed (changes requested):
I like where this is going, but I'm not loving the way the model integrates with the algorithms. 

What we should have is basically: 

 * There is (or are) virtual wallet(s).
 * Each virtual wallets is in some state.
 * Virtual wallet's state can be changed. 

Virtual wallets take precedence over platform [=holders=] when in automation.

timcappalli commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+          <tbody>
+            <tr>
+              <td>
+                Virtual Wallet Support
```
```suggestion
                Virtual Credential Manager Support
```

marcoscaceres replied:
We haven't really used "wallet" formally... we've stuck to holder... so I wonder if this should be "virtual holder"

TallTed replied:
I think "virtual (whatever)" is not accurate and will be misunderstood.

Better might be "software (whatever)" (even if it's implemented on hardware, it's still software doing the lifting).

Rather than `Virtual Credential Manager Support`, I would suggest `Support for Software Credential Managers` or perhaps `Support for Agentic Credential Managers`.

timcappalli commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+        <p>
+          When <a data-cite="webdriver#dfn-matching-capabilities">matching
+          capabilities</a>, the extension-specific steps to match
+          <code>digital-credentials:virtualWallet</code> with
```
```suggestion
          <code>digital-credentials:virtualCredentialManager</code> with
```

timcappalli commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+            </li>
+          </ol>
+        </section>
+        <section id="set-wallet-to-wait">
```
```suggestion
        <section id="set-cm-to-wait">
```

timcappalli commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+            The [=remote end steps=] are:
+          </p>
+          <ol class="algorithm">
+            <li>Set the session's [=next wallet behavior=] to <code>{ "mode":
```
```suggestion
            <li>Set the session's [=next credential manager behavior=] to <code>{ "mode":
```

TallTed commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+          </li>
+        </ol>
+      </section>
+      <section id="virtual-wallet-commands">
```
I don't know how much acronym avoidance matters in fragment identifiers, but it still might be better to abbreviate somewhat less, perhaps as —
```suggestion
      <section id="virtual-cred-mgr-commands">
```

mohamedamir commented on index.html:
```diff
@@ -917,6 +933,22 @@ <h3>
+          </li>
+        </ol>
+      </li>
+      <li>Otherwise (if [=next wallet behavior=] is not set), proceed with the
```
I will revert all changes to the algorithm as discussed in another comment thread.

mohamedamir commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+        </section>
+        <section id="set-wallet-to-decline">
+          <h4>
+            <dfn>Set Wallet to Decline</dfn>
```
YEs, but it's to simulate a user "declining" the flow.
I feel "decline" is more natural.
Let me know if you feel strongly.

mohamedamir committed:
Reverting updates to the algorithms

mohamedamir committed:
Reverting updates to the algorithms

mohamedamir committed:
tidy

mohamedamir committed:
tidy

marcoscaceres commented on index.html:
```diff
@@ -2267,7 +2141,265 @@ <h2>
+        For the purposes of user agent automation and web application testing,
+        this document defines a number of [[WebDriver]] [=extension commands=].
+      </p>
+      <section id="webdriver-extension-capability">
```
Nit: no need for section here.

marcoscaceres commented on index.html:
```diff
@@ -2267,7 +2141,265 @@ <h2>
+      </section>
+      <section id="virtual-wallet-commands">
+        <h3>
+          <code>Virtual Wallet Commands</code>
```
Nit: why wrap in code?

marcoscaceres commented on index.html:
```diff
@@ -2267,7 +2141,265 @@ <h2>
+                </th>
+              </tr>
+            </thead>
+            <tbody>
```
Nit: tbody is redundant here... gets added "for free".

mohamedamir committed:
Address comments

mohamedamir committed:
Address comments

mohamedamir committed:
resurrect mistakelnly deleted sections

mohamedamir committed:
resurrect mistakelnly deleted sections

mohamedamir committed:
Typo

mohamedamir committed:
Typo

mohamedamir commented on index.html:
```diff
@@ -2141,6 +2173,268 @@ <h2>
+            The [=remote end steps=] are:
+          </p>
+          <ol class="algorithm">
+            <li>If <code>parameters</code> is not a JSON <a data-cite=
```
@sadym-chromium 
Hi Maksim,

Could you chime in here?
Also, a review would be really appreciated!

Thank you!

whimboo commented on index.html:
```diff
@@ -2136,6 +2136,248 @@ <h2>
     </section>
+    <section data-cite="webdriver">
+      <h2>
+        User Agent Automation
```
If you consider to add support for WebDriver BiDi at a later point it might be good to already put the classic support into its subsection. Otherwise it's unclear which specification this PR is related (unless you check the links in detail).

mohamedamir committed:
Avoid data-cite and reference webdrive

mohamedamir committed:
Avoid data-cite and reference webdrive

mohamedamir committed:
restructure automation section to distinguish WebDriver Classic and BiDi

- Move current automation logic into a new "WebDriver Classic" subsection.
- Add a placeholder subsection for "WebDriver BiDi" to improve clarity.

mohamedamir committed:
restructure automation section to distinguish WebDriver Classic and BiDi

- Move current automation logic into a new "WebDriver Classic" subsection.
- Add a placeholder subsection for "WebDriver BiDi" to improve clarity.

mohamedamir committed:
Link to the algorithms, not to the methods.

mohamedamir committed:
Link to the algorithms, not to the methods.

mohamedamir committed:
refactor: Migrate WebDriver automation section from Classic to BiDi, updating capability keys and command structures.

mohamedamir committed:
refactor: Migrate WebDriver automation section from Classic to BiDi, updating capability keys and command structures.

mohamedamir committed:
docs: Remove Digital Credentials WebDriver BiDi Extension Capability section.

mohamedamir committed:
docs: Remove Digital Credentials WebDriver BiDi Extension Capability section.

mohamedamir committed:
refactor: adjust heading levels for virtual wallet commands and their sub-sections.

mohamedamir committed:
refactor: adjust heading levels for virtual wallet commands and their sub-sections.

sadym-chromium commented on index.html:
```diff
@@ -2136,6 +2136,166 @@ <h2>
+          </thead>
+          <tr>
+            <td>
+              `digitalCredentials.setResponse`
```
Is the scope of the command global for the whole browser? Is it aligned with implementation? Other option would be to make it more flexible and set per top-level browsing context (page), user context (browser context in CDP terms) or global.

mohamedamir replied:
This has been addressed!

mohamedamir committed:
Refactor virtual wallet testing by replacing `next wallet behavior` and multiple WebDriver-BiDi commands with `active virtual wallet behavior` and a unified `setVirtualWalletBehavior` command.

mohamedamir committed:
Refactor virtual wallet testing by replacing `next wallet behavior` and multiple WebDriver-BiDi commands with `active virtual wallet behavior` and a unified `setVirtualWalletBehavior` command.

mohamedamir committed:
docs: Update digital credential method references and add serialization and error handling for the virtual wallet behavior command.

mohamedamir committed:
docs: Update digital credential method references and add serialization and error handling for the virtual wallet behavior command.

whimboo commented on index.html:
```diff
@@ -2453,6 +2507,107 @@ <h2>
+)
+        </pre>
+            <p>
+              <strong>Return Type:</strong> `EmptyResult`
```
Please make sure to also tag the return type as CDDL Also each CDDL definition needs the `local-cddl` or `remote-cddl` attribute set. You can check the BiDi or another specification (which implements BiDi commands) in how to get this done.

mohamedamir committed:
fix: updated the `response` type from object to text, and refactored return type display using CDDL modules.

mohamedamir committed:
fix: updated the `response` type from object to text, and refactored return type display using CDDL modules.

mohamedamir committed:
refactor: Extract virtual wallet behavior handling into a dedicated algorithm.

mohamedamir committed:
refactor: Extract virtual wallet behavior handling into a dedicated algorithm.

mohamedamir committed:
refactor: Update `response` parameter type definition and adjust its serialization and parsing logic for virtual wallet actions.

mohamedamir committed:
refactor: Update `response` parameter type definition and adjust its serialization and parsing logic for virtual wallet actions.

mohamedamir committed:
Reorder virtual wallet handling to occur after credential request preparation.

mohamedamir committed:
Reorder virtual wallet handling to occur after credential request preparation.

TallTed reviewed (changes requested):
Small stuff

whimboo commented on index.html:
```diff
@@ -2453,6 +2463,163 @@ <h2>
+  ? response: { * text =&gt; any },
+}
+    </pre>
+          <ul>
```
Lets make sure those entries are all related to `action` within `SetVirtualWalletBehaviorParameters`. Maybe add as well a sentence first which starts with: `The <code>SetVirtualWalletBehaviorParameters</code> type represents ...`.

mohamedamir replied:
Thank you both very much.!
@whimboo Thanks a lot for all the guidance and support through out this PR!
Much appreciated!

whimboo reviewed (changes requested):
Thanks. I've a couple more questions and notes regarding the WebDriver specific section.

mohamedamir committed:
Apply suggestions from code review

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>
Co-authored-by: Henrik Skupin <mail@hskupin.info>

mohamedamir committed:
Apply suggestions from code review

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>
Co-authored-by: Henrik Skupin <mail@hskupin.info>

mohamedamir committed:
docs: reformat CDDL blocks and reorder list items in digital credentials specification

mohamedamir committed:
docs: reformat CDDL blocks and reorder list items in digital credentials specification

mohamedamir committed:
docs: add documentation for digitalCredentials.VirtualWalletAction type to index.html

mohamedamir committed:
docs: add documentation for digitalCredentials.VirtualWalletAction type to index.html

mohamedamir committed:
docs: fix formatting of VirtualWalletAction type description in index.html

mohamedamir committed:
docs: fix formatting of VirtualWalletAction type description in index.html

mohamedamir committed:
refactor: simplify command parameter field access in algorithm steps

mohamedamir committed:
refactor: simplify command parameter field access in algorithm steps

mohamedamir committed:
feat: add error handling for non-null responses in active virtual wallet behavior

mohamedamir committed:
feat: add error handling for non-null responses in active virtual wallet behavior

TallTed commented on index.html:
```diff
@@ -1395,6 +1395,11 @@ <h3>
+      </li>
+      <li>If |handled| is `true`, return |promise|.
+      </li>
       <li>Return |promise|.
```
What's the point of this `If` test, given that both `|handled| is true` and *all other values of `|handled|`* (thus, in aggregate, *all values of `|handled|`*) result in `return |promise|`?

mohamedamir replied:
The algorithm isn't complete yet

It will be something like 
```
      <li>[=credential request coordinator/Prepare credential requests=] with
      |document|, |requests|, |promise|, and |signal|.
      </li>
      <li>Let |handled| be the result of running [=handle virtual wallet
      behavior=] given |promise| and |global|.
      </li>
      <li>If |handled| is `true`, return |promise|.
      </li>
      <li>[=credential request coordinator/Present credential requests=] with
      |document|, |requests|, |promise|, and |signal|.
      </li>
      <li>Return |promise|.
```

once https://github.com/w3c-fedid/digital-credentials/pull/419 is merged.!

marcoscaceres commented:
I did a systematic review of the testing framework against all the algorithms (including the new "present the credential request" algorithm from #419). Here's what I found:

## What the framework CAN test (great coverage)

The `"wait"` action is especially useful — it enables testing:
- **Concurrent requests** — `"wait"` + second `get()` call to verify the error
- **Abort signal during presentation** — `"wait"` + abort the signal
- **Document becoming inactive** — `"wait"` + navigate away

And `"respond"` covers the happy path well.

## Issues I'd like to discuss

### 1. `"decline"` rejects with `NotAllowedError` — should be `AbortError`

The `"decline"` action currently rejects with `NotAllowedError`. But we agreed in the [#419 discussion](https://github.com/w3c-fedid/digital-credentials/pull/419#discussion_r2561523770) that user cancellation should reject with `AbortError` — distinct from platform permission denial (`NotAllowedError`). The "present the credential request" algorithm (step 5) reflects this:

> If the user cancels the operation or no credential was selected: Let |error| be a newly created `"AbortError"` DOMException.

Since `"decline"` simulates the user dismissing the chooser, it should reject with `AbortError` to match the spec. `NotAllowedError` is reserved for when the *platform* denies permission (step 6a).

### 2. `"respond"` doesn't set `protocol`

The `"respond"` action creates a `DigitalCredential` with `.data` but doesn't set `.protocol`. The spec requires both attributes (present step 7c-e). Tests can't verify that the correct protocol enum value is returned.

Should the `"respond"` action accept an optional `protocol` parameter? Or should it default to the first valid protocol from the request?

### 3. No way to test platform error mapping

The "present" algorithm maps 4 types of platform errors (step 6):
- Permission denied → `NotAllowedError`
- Malformed request data → `TypeError`  
- Already in progress → `InvalidStateError`
- Catch-all → `OperationError`

With only respond/decline/wait, none of these error conditions are testable. Would it be worth adding an `"error"` action that accepts an error type parameter to simulate platform-level failures?

### 4. No way to test response parsing failures

Because `"respond"` pre-validates the response as a JSON object (`{ * text => any }`), there's no way to test:
- Holder returning invalid JSON (present step 7a → parse error)
- Holder returning valid JSON that isn't an object, e.g. `"hello"` or `42` (present step 7b → `TypeError`)

A `"respond-raw"` action accepting an arbitrary string could cover these paths.

## Summary

| Framework action | What it tests | Gap |
|-----------------|---------------|-----|
| `"respond"` | Happy path | Missing `protocol`; can't test bad responses |
| `"decline"` | User cancellation | Wrong error type (`NotAllowedError` vs `AbortError`) |
| `"wait"` | Concurrent requests, abort, timeouts | None — works well |
| (missing) | Platform errors | No error injection capability |
| (missing) | Response parsing failures | No raw response injection |

Happy to help iterate on this. The framework is a great foundation — these are refinements to get full algorithm coverage.

marcoscaceres reviewed (changes requested):
See https://github.com/w3c-fedid/digital-credentials/pull/381#issuecomment-4221822734 ... there's still some gaps. 

We can split those into issues or address them here. I don't mind.

mohamedamir committed:
fix: remove wrong text during rebase

mohamedamir committed:
fix: remove wrong text during rebase

mohamedamir commented on index.html:
```diff
@@ -1452,6 +1457,11 @@ <h3>
+      </li>
+      <li>If |handled| is `true`, return |promise|.
+      </li>
       <li>Return |promise|.
```
Similar to above!

mohamedamir committed:
feat: require protocol in respond action for virtual wallet behavior

mohamedamir committed:
feat: require protocol in respond action for virtual wallet behavior

mohamedamir commented:
Thanks @marcoscaceres for the feedback

> ### 1. `"decline"` rejects with `NotAllowedError` — should be `AbortError`

That's not correct, we agreed to return `NotAllowedError` for user cancellation, and keep `AbortError` for programmatic Abort from the developer.

> ### 2. `"respond"` doesn't set `protocol`
Great find, fixed!

> ### 3. No way to test platform error mapping
I am not sure if this is something we would like to add given that platform error aren't standardized.
WDYT?

> ### 4. No way to test response parsing failures
I filed https://github.com/w3c-fedid/digital-credentials/issues/498 to track this issue, since I am not sure it is worth the complexity at the moment.

marcoscaceres commented:
CDDL suport inbound! https://github.com/speced/respec/pull/5201 😎

marcoscaceres commented:
> I am not sure if this is something we would like to add given that platform error aren't standardized.
> WDYT?

See the "present" branch... I've put them into classes of errors, so maybe we could? We could even formalize the classes of platform errors for this purpose and link to them from here. That would be kinda nice.

marcoscaceres commented:
> That's not correct, we agreed to return NotAllowedError for user cancellation, and keep AbortError for programmatic Abort from the developer.

Yeah, sorry... I think I was looking at the wrong version of the spec.

mohamedamir commented:
> > I am not sure if this is something we would like to add given that platform error aren't standardized.
> > WDYT?
> 
> See the "present" branch... I've put them into classes of errors, so maybe we could? We could even formalize the classes of platform errors for this purpose and link to them from here. That would be kinda nice.

Fair enough, I have submitted https://github.com/w3c-fedid/digital-credentials/issues/500 to track this.
We can revisit once the present part is merged!

marcoscaceres commented:
Thanks for working on this. WebDriver automation is critical for testability, and the overall approach (virtual wallet behavior rather than full virtual wallets) is the right one. We've been evaluating this from a WebKit implementation perspective, and I want to flag several issues we've found. Grouped by category:

## Algorithm ordering

The "handle virtual wallet behavior" hook is inserted *after* the "Prepare credential requests" step. In WebKit, "Prepare" validates requests, transitions the coordinator to a "Presenting" state, and calls `showDigitalCredentialsPicker()`, which sends IPC to the UIProcess to present a system sheet. By the time the virtual wallet check runs, the picker is already being shown.

This means:

- The virtual wallet would resolve the promise while the coordinator's state machine is still in "Presenting" state, bypassing the normal resolution flow.
- The coordinator's `dismissDigitalCredentialsPicker` is never called, potentially leaving the UIProcess picker alive.

The virtual wallet check needs to run before the presentation flow begins.

## Scoping and lifecycle

The spec says "Set the **user agent's** active virtual wallet behavior." This is a UA-global singleton. Other WebDriver automation APIs scope their state differently: FedCM commands use `/session/{session id}/fedcm/...` (session-scoped), and WebAuthn virtual authenticator commands take a `browsingContextHandle` parameter (context-scoped).

UA-global scoping has two consequences:

1. **Test isolation is broken.** A behavior set in one test leaks into subsequent tests unless explicitly overwritten. There is no `clearVirtualWalletBehavior` command and no specified cleanup when the WebDriver session ends. Compare with WebAuthn, which provides `removeVirtualAuthenticator` and `removeAllVirtualAuthenticatorCredentials`.

2. **Cross-context interference during automation.** If multiple tabs or sessions are active, a page in one context could observe the virtual wallet behavior configured for a different context.

## Automation security boundary

The `handle virtual wallet behavior` algorithm runs on every credential request, including in non-automated browsing. It checks whether the "active virtual wallet behavior" is null, but does not verify that the user agent is under automation. While `setVirtualWalletBehavior` is gated to WebDriver BiDi (content JS cannot invoke it), adding an explicit "if not under automation, return false" guard as step 0 would provide defense in depth.

## BiDi/Classic protocol mismatch

We also reviewed the linked WPT PR ([web-platform-tests/wpt#58587](https://github.com/web-platform-tests/wpt/pull/58587)) to understand how the spec's commands would be exercised in practice. We realize the WPT implementation may not have been updated to match the current spec text yet, but flagging the current gaps so they're tracked:

This PR defines only BiDi commands (CDDL types like `digitalCredentials.SetVirtualWalletBehaviorParameters`), but the WPT implementation currently uses WebDriver Classic HTTP endpoints: `POST digital-credentials/response`, `POST digital-credentials/response_wait`, and `POST digital-credentials/response_decline`. These Classic endpoints are not defined in this PR or the WebDriver specification. The WPT tests will need to align with whichever wire protocol the spec ultimately defines.

The WPT testdriver API also currently diverges from the spec's command shape in a couple of ways that will need reconciling:

- `set_virtual_wallet_response(data)` passes a plain string as `data`, but the spec's CDDL requires `response` to be `{ * text => any }` (a map).
- The testdriver omits the `protocol` parameter, which the spec requires to be non-null when action is "respond" (step 4.1 would return an error).

## Interoperability

The module is OPTIONAL, but the WPT tests have no feature detection or metadata gating. They will permanently fail in browsers that don't implement the module. Gecko has no implementation commitment listed in this PR.

The tests also hard-code `openid4vp-v1-unsigned` and `openid4vci` as protocol values. Since the virtual wallet bypasses all protocol validation and returns whatever was configured, the tests pass without exercising any browser's actual protocol-handling code paths.

## Adversarial testing capability

For security testing, we'd want this API to support two attack directions:

**Wallet → RP/content process:** Can the virtual wallet return data that tests the content process's resilience? The current API allows arbitrary response shapes, which is good. But there is no timing control (the "wait" action is binary, with no configurable delay for testing race conditions), no mechanism for sequenced responses (e.g., "decline the first request, respond to the second"), and no way to configure different behaviors for `create()` vs `get()`.

**RP → wallet/UA:** The web page already controls request data, but the virtual wallet ignores it and returns canned responses. There is no way for a test to observe what the UA forwarded to the virtual wallet (e.g., whether `topOrigin` and `documentOrigin` were correctly included, whether request validation occurred). Compare with WebAuthn's `getVirtualAuthenticatorCredentials`, which provides this kind of introspection.

marcoscaceres committed:
chore: use ReSpec v36 CDDL inline references (#502)

Replace plain-text CDDL type references with {^ ^} inline syntax
for auto-linking to CDDL definitions. Add CDDL index section.

No normative text changes.

marcoscaceres committed:
chore: use ReSpec v36 CDDL inline references (#502)

Replace plain-text CDDL type references with {^ ^} inline syntax
for auto-linking to CDDL definitions. Add CDDL index section.

No normative text changes.

mohamedamir committed:
Address WebDriver automation comments regarding scoping, ordering, and security

mohamedamir committed:
Address WebDriver automation comments regarding scoping, ordering, and security

mohamedamir commented:
Hi @marcoscaceres,

Thanks for the detailed review and for flagging these issues! I have addressed your feedback in the latest commits. Here is a summary of how I handled each point:

1. **Algorithm ordering**: I moved the `handle virtual wallet behavior` hook inside the `Prepare credential requests` algorithm, specifically **after** the validation steps but **before** `Present the credential request`. This ensures that request validation still occurs while avoiding triggering the UI presentation flow in automated tests.
2. **Scoping and lifecycle**: I have updated the spec to associate the `active virtual wallet behavior` with the **`WebDriver session`** instead of the global `user agent`. This should prevent cross-session interference. I also added a `"clear"` action to `VirtualWalletAction` to allow tests to explicitly reset the behavior, addressing the lack of a cleanup mechanism.
3. **Automation security boundary**: I added an explicit check as the first step in `Handle Virtual Wallet Behavior`: *"If the user agent is not under automation, return false."*
4. **BiDi/Classic protocol mismatch**:  I am working on updating the WPT tests to use WebDriver BiDi to align with this spec.

Please let me know if this addresses your concerns or if you have further feedback!

TallTed commented:
There appear to have been a large number of changes since [my last review](https://github.com/w3c-fedid/digital-credentials/pull/381#pullrequestreview-4022541675). It seems likely that I don't need to [re-review every line of this PR](https://github.com/w3c-fedid/digital-credentials/pull/381#event-24803368395), but rather only those recent changes. However, GitHub is refusing to provide a link to "all changes since I last viewed". Perhaps you can quickly build such a link?

marcoscaceres commented:
Hi @TallTed,

Mohamed's most recent commit (`ba531c5`, April 23) addresses feedback I gave earlier that day. Here's what changed in the spec language:

**1. Algorithm ordering (moved the virtual wallet hook earlier)**
The "handle virtual wallet behavior" steps were removed from both `get()` and `create()` (where they ran *after* "Prepare credential requests") and moved *into* the "Prepare credential requests" algorithm itself, before the presentation flow begins. This prevents the virtual wallet from resolving a promise while the UI picker is already being shown.

**2. Scoping: session instead of global**
All references to "the user agent's active virtual wallet behavior" were changed to "the WebDriver session's active virtual wallet behavior." This scopes the state to individual sessions rather than being UA-global, preventing cross-session test interference.

**3. New "clear" action**
A `"clear"` value was added to `VirtualWalletAction`. The `setVirtualWalletBehavior` algorithm now checks for `"clear"` and sets the session's active behavior to `null`, giving tests an explicit cleanup mechanism.

**4. Automation security guard**
A new first step was added to the "handle virtual wallet behavior" algorithm: "If the user agent is not under automation, return false." This is a defense-in-depth check so the virtual wallet path can never run outside of WebDriver automation.

Those are the substantive language changes. Some minor reformatting of the `VirtualWalletAction` description list also happened (line wrapping), but no wording changes there.

**Suggested focus areas for your editorial eye:**

- **The `VirtualWalletAction` descriptions** ("decline", "respond", "wait", "clear" bullet list). These were rewrapped and reformatted. Worth checking that they're parallel in structure, consistent in voice, and clear to a first-time reader.
- **The "handle virtual wallet behavior" algorithm steps.** These were relocated and reworded ("user agent's" became "WebDriver session's"). Worth checking that the new phrasing reads naturally and that subject/object references are unambiguous.
- **The `"clear"` action description** is noticeably shorter than the other three. It might benefit from a bit more explanatory text to match the level of detail in its siblings.
- **Consistency of "return false" / "return true"** language across the handle algorithm. The new automation guard and the existing steps should use a consistent style.

Hope that helps!

mohamedamir committed:
Update setVirtualWalletBehavior for WebDriver BiDi frame scoping

mohamedamir committed:
Update setVirtualWalletBehavior for WebDriver BiDi frame scoping

mohamedamir commented on index.html:
```diff
@@ -1680,8 +1685,8 @@ <h3>
       </li>
-      <li>Let |promise| be [=a new promise=] in the [=relevant realm=] of
-      [=this=].
+      <li>Let |promise| be [=a new promise=].
```
```suggestion
      <li>Let |promise| be [=a new promise=] in the [=relevant realm=] of
      [=this=].
```

mohamedamir committed:
Apply suggestions from code review

Co-authored-by: Mohamed Amir Yosef <mohamed.amir@gmail.com>

mohamedamir commented on index.html:
```diff
@@ -1682,6 +1687,7 @@ <h3>
       </li>
       <li>Let |promise| be [=a new promise=] in the [=relevant realm=] of
       [=this=].
+
```
```suggestion
```

mohamedamir committed:
Apply suggestions from code review

Co-authored-by: Mohamed Amir Yosef <mohamed.amir@gmail.com>

marcoscaceres committed:
fix: link Infra algorithm + tidy formatting

- Link "parsing JSON into a JavaScript value" to the Infra spec's
  "parse a JSON string to a JavaScript value" algorithm.
- Remove stray blank line in promise creation step.
- Run tidy (80-char line wrapping).

marcoscaceres committed:
fix: validate protocol enum + clarify wait behavior

- Add validation step in setVirtualWalletBehavior: the protocol
  parameter must be a valid DigitalCredentialProtocol enum value
  when action is "respond".
- Add notes clarifying that unsettled "wait" promises must be
  rejected with AbortError when the session ends.
- Add note clarifying that concurrent "wait" requests for the
  same context must abort the previous request.

marcoscaceres committed:
fix: remove redundant serialization in handle virtual wallet behavior

The `|response|` stored in the behavior tuple is already a JSON string
(serialized in setVirtualWalletBehavior step 5c). Re-serializing it in
the handle algorithm would double-encode it. Parse it directly instead.

marcoscaceres commented on index.html:
```diff
@@ -2705,8 +2710,235 @@ <h2>
+          </li>
+          <li>Let (|action|, |protocol|, |response|) be |behavior|.
+          </li>
+          <li>If |action| is `"wait"`, return `true`.
```
**Context destruction during "wait"**

When the `"wait"` action leaves a promise unsettled, the spec says the session-end MUST reject it (covered in the note below). But what about browsing context destruction? If the tab/frame that initiated the credential request is closed while the promise is pending, the spec doesn't explicitly require the UA to reject it.

Suggestion: add a normative requirement like: "If the browsing context associated with a pending `"wait"` promise is destroyed, the user agent MUST reject the promise with an `"AbortError"` DOMException."

mohamedamir replied:
Done!

marcoscaceres commented on index.html:
```diff
@@ -2705,8 +2710,235 @@ <h2>
+          When the [=WebDriver session=] ends, the [=user agent=] MUST
+          [=reject=] any unsettled {{Promise}} objects that were left pending
+          by the `"wait"` action with an {{"AbortError"}} {{DOMException}}.
+        </p>
```
**Individual request validation failures (existing spec text, but relevant to this section)**

The "validate credential requests" algorithm (step 8 in the existing spec) says "If validation fails, throw an appropriate exception." This makes a single bad request kill the entire operation. Some UAs (WebKit) skip individual invalid requests and log a console warning instead, which is arguably better UX (one bad request shouldn't block valid ones).

The spec might want to allow this: "The user agent MAY skip individual requests that fail protocol-specific validation, logging a developer-visible warning, rather than throwing." The empty-result case is already handled by step 7 of "prepare credential requests" (TypeError if validatedRequests is empty).

mohamedamir replied:
This doesn't belong to this PR, does it?
This is related to the "validate credential requests" algorithm not the user agent automation.

Am I overlooking something?

marcoscaceres replied:
no, oops.. fixing now...

marcoscaceres committed:
Revert "fix: remove redundant serialization in handle virtual wallet behavior"

This reverts commit fc04576db2ca29baf74ea95d6e38fb513f049785.

marcoscaceres committed:
Merge branch 'main' into user-agent-automation

marcoscaceres reviewed (approved):
A few comments after implementing, but looks good. Excellent work @mohamedamir!

mohamedamir committed:
docs: specify abort behavior for pending wait promises when browsing context is destroyed

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,243 @@ <h2>
+          When the [=WebDriver session=] ends, the [=user agent=] MUST
+          [=reject=] any unsettled {{Promise}} objects that were left pending
+          by the `"wait"` action with an {{"AbortError"}} {{DOMException}}.
+        </p>
```
Sorry, this was supposed to be advisory:  

```suggestion
        <p class="note">
          When the [=WebDriver session=] ends, the [=user agent=]
          [=rejects=] unsettled {{Promise}} objects that were left pending
          by the `"wait"` action.
        </p>
```

It's the same as destroying the the frame ... so it just behaves the same way as the document is "no longer fully active".

marcoscaceres committed:
Apply suggestion from @marcoscaceres

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,243 @@ <h2>
+          If the [=browsing context=] associated with a pending `"wait"`
+          promise is destroyed, the [=user agent=] MUST [=reject=] the promise
+          with an {{"AbortError"}} {{DOMException}}.
+        </p>
```
```suggestion
```

marcoscaceres committed:
Apply suggestion from @marcoscaceres

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
         MUST be operable in a device-independent manner.
       </p>
     </section>
+    <section data-cite="webdriver webdriver-bidi">
+      <h2>
+        Automated Testing
+      </h2>
+      <p>
+        For the purposes of user-agent automation and application testing, this
+        document defines extension modules for the [[[webdriver-bidi]]]
+        specification.
+      </p>
+      <p>
+        It is OPTIONAL for a [=user agent=] to support them.
+      </p>
```
```suggestion
      <p>
        For the purposes of user-agent automation and application testing, this
        document defines extension modules for the [[[webdriver-bidi]]]
        specification. It is OPTIONAL for a [=user agent=] to support them.
      </p>
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+      <p>
+        It is OPTIONAL for a [=user agent=] to support them.
+      </p>
+      <section>
```
Nit: we can delete this `<section>` (ReSpec adds section elements automagically)

mohamedamir replied:
Done!

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+          {{Credential/[[Create]](origin,options, sameOriginWithAncestors)}}
+          calls.
+        </p>
+        <section>
```
As above.

mohamedamir replied:
Done!

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+              }
+          </pre>
+          <p>
+            The {^ digitalCredentials.VirtualWalletAction ^} type represents
```
```suggestion
            The {^digitalCredentials.VirtualWalletAction^} type represents
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+                ? protocol: text,
+                ? response: { * text =&gt; any },
+              }
+          </pre>
```
Nit.. whitespace. 

```suggestion
          <pre class="cddl" data-cddl-module="remote-cddl">
            digitalCredentials.VirtualWalletAction = "decline" / "respond" / "wait" / "clear"
  
            digitalCredentials.SetVirtualWalletBehaviorParameters = {
              action: digitalCredentials.VirtualWalletAction,
              ? context: text,
              ? protocol: text,
              ? response: { * text =&gt; any },
            }
          </pre>
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+            the different types of virtual wallet actions.
+          </p>
+          <ul>
+            <li>{^ digitalCredentials.VirtualWalletAction/"decline" ^}: The
```
```suggestion
            <li>{^digitalCredentials.VirtualWalletAction/"decline"^}: The
```

mohamedamir committed:
chore: update WebDriver session references to use proper spec linking syntax

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+            virtual wallet simulates a user cancellation or rejection, aborting
+            the request.
+            </li>
+            <li>{^ digitalCredentials.VirtualWalletAction/"respond" ^}: The
```
```suggestion
            <li>{^digitalCredentials.VirtualWalletAction/"respond"^}: The
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+            virtual wallet simulates a successful user interaction and returns
+            the predefined credential response.
+            </li>
+            <li>{^ digitalCredentials.VirtualWalletAction/"wait" ^}: The
```
```suggestion
            <li>{^digitalCredentials.VirtualWalletAction/"wait"^}: The
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+            leaving the active promise unsettled to test timeouts or concurrent
+            request handling.
+            </li>
+            <li>{^ digitalCredentials.VirtualWalletAction/"clear" ^}: Clears
```
```suggestion
            <li>{^digitalCredentials.VirtualWalletAction/"clear"^}: Clears
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+            The {^ digitalCredentials.VirtualWalletAction ^} type represents
+            the different types of virtual wallet actions.
+          </p>
+          <ul>
```
This looks like it should be a definition list that defines each thing, instead of a ul. 

```suggestion
          <dl>
```

marcoscaceres commented on index.html:
```diff
@@ -2746,8 +2751,227 @@ <h2>
+            <h5>
+              The `digitalCredentials.setVirtualWalletBehavior` Command
+            </h5>
+            <p>
```
Same with these `<p>` strong combos ... these should be in a `dl`

marcoscaceres reviewed (commented):
Some nits... getting copilot on them.

Copilot committed:
editorial: apply automation section nits and markup cleanup

Agent-Logs-Url: https://github.com/w3c-fedid/digital-credentials/sessions/fea7d8bc-83ce-4fec-9648-bf343c35ff8a

Co-authored-by: marcoscaceres <870154+marcoscaceres@users.noreply.github.com>

mohamedamir committed:
refactor: update virtual wallet behavior terminology to use modules/session instead of WebDriver session

mohamedamir committed:
refactor: improve formatting and remove redundant section nesting in digital credentials specification

TallTed commented on index.html:
```diff
@@ -2767,9 +2767,8 @@ <h3>
-        sameOriginWithAncestors)}} and
-        {{Credential/[[Create]](origin,options, sameOriginWithAncestors)}}
-        calls.
+        sameOriginWithAncestors)}} and {{Credential/[[Create]](origin,options,
```
for consistency
```suggestion
        sameOriginWithAncestors)}}, and {{Credential/[[Create]](origin, options,
```

TallTed commented on index.html:
```diff
@@ -2746,8 +2751,230 @@ <h2>
+        Automated Testing
+      </h2>
+      <p>
+        For the purposes of user-agent automation and application testing, this
```
```suggestion
        For purposes of user-agent automation and application testing, this
```

mohamedamir committed:
Apply suggestions from from Ted

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

TallTed commented on index.html:
```diff
@@ -1680,8 +1685,8 @@ <h3>
       </li>
-      <li>Let |promise| be [=a new promise=] in the [=relevant realm=] of
-      [=this=].
+      <li>Let |promise : Promise| be [=a new promise=] in the [=relevant
```
This appears to be the only instance of ` : ` (space-colon-space) in this document. I haven't checked the other documents in the repo so I don't know whether it's the only instance of ` : ` in the repo. It appears to be a typo. Please confirm or deny... :-)

marcoscaceres replied:
ReSpec’s inline-variable parser doesn’t mind, but we can quickly fit it. @copilot can you send a PR for consistency fixing this?
